extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,470 @@
+ *     interval expires.
+ * (5) Once the interval expires, the element will enter the REFRESHED state.
+ *     The element will remain in this state until reset() is called on the
+ *     element, at which point it will return to the INITIAL state.
+ *
+ * @enum {number}
+ */
+const RefreshLifecycleState = {
+  /**
+   * Element has been registered, but not yet seen on screen.
+   */
+  INITIAL: 0,
+
+  /**
+   * The element has appeared in the viewport, but not yet for the required
+   * duration.
+   */
+  VIEW_PENDING: 1,
+
+  /**
+   * The element has been in the viewport for the required duration; the
+   * refresh interval for the element has begun.
+   */
+  REFRESH_PENDING: 2,
+
+  /**
+   * The refresh interval has elapsed; the element is in the process of being
+   * refreshed.
+   */
+  REFRESHED: 'refreshed',
+};
+
+/**
+ * Returns the singleton instance of the RefreshManager from the window object,
+ * or creates and returns one, if one has not yet been constructed.
+ *
+ * @param {!Window} win
+ * @return {!RefreshManager}
+ */
+export function getRefreshManagerFor(win) {
+  return win[NAME_IN_WINDOW] ||
+      (win[NAME_IN_WINDOW] = new RefreshManager(win));
+}
+
+/** Visible for testing */
+export function resetRefreshManagerFor(win) {
+  getRefreshManagerFor(win).resetManager();
+  win[NAME_IN_WINDOW] = null;
+}
+
+export class RefreshManager {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+
+    /** @const @private {!Window} */
+    this.win_ = win;
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = timerFor(win);
+
+    /**
+     * An object containing the IntersectionObservers used to monitor elements.
+     * Each IO is configured to a different threshold, and all elements that
+     * share the same minOnScreenPixelRatioThreshold will be monitored by the
+     * same IO.
+     *
+     * @private {!Object<string, (IntersectionObserver|!IntersectionObserverPolyfill)>}
+     */
+    this.intersectionObservers_ = {};
+
+    /**
+     * A an object containing the RefreshState of all elements currently being
+     * monitored, indexed by a unique ID stored as data-amp-a4a-refresh-id on
+     * the element.
+     *
+     * @private {!Object<string, RegisteredElementWrapper>}
+     */
+    this.registeredElementWrappers_ = {};
+
+    /**
+     * Used to distinguish elements in this.intersectionObservers_. This is
+     * incremented with each newly registered element.
+     *
+     * @private {number}
+     */
+    this.elementReferenceId_ = 0;
+  }
+
+  /**
+   * Returns an IntersectionObserver configured to the given threshold, creating
+   * one if one does not yet exist.
+   *
+   * @param {string|number} threshold
+   * @return {(IntersectionObserver|!IntersectionObserverPolyfill)}
+   */
+  getIntersectionObserverWithThreshold_(threshold) {
+    threshold = String(threshold);
+    this.intersectionObservers_[threshold] =
```

Comment:
This is nested a bit deeply. Maybe just `if (!this.intersectionObservers_[threshold])`?